### PR TITLE
Add blob container soft delete rule

### LINKED ISSUE #1671: Enabled storage container soft delete
# Rule request

## Suggested rule change

Like blob storage, containers should also have soft delete enabled. Days should be defined however the number of days is not important.

See https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.Storage.SoftDelete/ for an example.

## Applies to the following

The rule applies to the following:

- Resource type: **Microsoft.Storage/storageAccounts**

## Additional context

https://learn.microsoft.com/en-gb/azure/architecture/framework/resiliency/design-requirements
[WAF](https://docs.microsoft.com/azure/architecture/framework/resiliency/data-management)
[Soft delete](https://learn.microsoft.com/azure/storage/blobs/soft-delete-container-overview)

### PR DESCRIPTION
**Pull request by pazdedav**
## PR Summary

Added a new rule (in PSH) that checks, if **soft delete** was enabled for a blob container.

Fixes #1671 

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

**Comment by pazdedav** [comment]
This is my first attempt to create a new rule, so I hope I understood what needs to be done:

- I added a new rule in `Azure.Storage.Rule.ps1`
- I create a documentation file - `Azure.Storage.ContainerSoftDelete.md` - with code samples for it (I picked **AZR-000289** as ID)
- I updated the test JSON ARM template - `Resources.Storage.Template.json` and `Resources.Storage.Parameters.json` with one new parameter, two new variables, and one new resource property
- I updated the unit test file - `Azure.Storage.Tests.ps1` file, added corresponding blocks
- I updated the `CHANGELOG-v1.md` file

Please provide your feedback, @BernieWhite. Cheers.

File: tests/PSRule.Rules.Azure.Tests/Azure.Storage.Tests.ps1 [hunk 1]
```diff
@@ -106,6 +106,33 @@ Describe 'Azure.Storage' -Tag Storage {
...
         }
 
+        It 'Azure.Storage.ContainerSoftDelete' {
+            $filteredResult = $result | Where-Object { $_.RuleName -eq 'Azure.Storage.ContainerSoftDelete' };
+
+            # Fail
+            $ruleResult = @($filteredResult | Where-Object { $_.Outcome -eq 'Fail' });
+            $ruleResult | Should -Not -BeNullOrEmpty;
+            $ruleResult.Length | Should -Be 3;
+            $ruleResult.TargetName | Should -BeIn 'storage-B', 'storage-C';
```
Comment (BernieWhite) [inline tests/PSRule.Rules.Azure.Tests/Azure.Storage.Tests.ps1:116]:
```suggestion
            $ruleResult.TargetName | Should -BeIn 'storage-B', 'storage-C', 'storage-A';
```

File: tests/PSRule.Rules.Azure.Tests/Azure.Storage.Tests.ps1 [hunk 2]
```diff
@@ -106,6 +106,33 @@ Describe 'Azure.Storage' -Tag Storage {
...
+
+            $ruleResult[0].Reason | Should -Not -BeNullOrEmpty;
+            $ruleResult[0].Reason | Should -BeExactly 'Path properties.containerDeleteRetentionPolicy.enabled: Is set to ''False''.';
+            $ruleResult[1].Reason | Should -Not -BeNullOrEmpty;
+            $ruleResult[1].Reason | Should -BeExactly 'A sub-resource of type ''Microsoft.Storage/storageAccounts/blobServices'' has not been specified.';
```
Comment (BernieWhite) [inline tests/PSRule.Rules.Azure.Tests/Azure.Storage.Tests.ps1:121]:
```suggestion
            $ruleResult[0].Reason | Should -BeExactly 'Path properties.containerDeleteRetentionPolicy.enabled: Does not exist.';
            $ruleResult[1].Reason | Should -Not -BeNullOrEmpty;
            $ruleResult[1].Reason | Should -BeExactly 'Path properties.containerDeleteRetentionPolicy.enabled: Does not exist.';
            $ruleResult[2].Reason | Should -Not -BeNullOrEmpty;
            $ruleResult[2].Reason | Should -BeExactly 'A sub-resource of type ''Microsoft.Storage/storageAccounts/blobServices'' has not been specified.';
```

File: tests/PSRule.Rules.Azure.Tests/Azure.Storage.Tests.ps1 [hunk 3]
```diff
@@ -106,6 +106,33 @@ Describe 'Azure.Storage' -Tag Storage {
...
+
+            # Pass
+            $ruleResult = @($filteredResult | Where-Object { $_.Outcome -eq 'Pass' });
+            $ruleResult | Should -Not -BeNullOrEmpty;
+            $ruleResult.Length | Should -Be 2;
+            $ruleResult.TargetName | Should -BeIn 'storage-A', 'storage-G';
```
Comment (BernieWhite) [inline tests/PSRule.Rules.Azure.Tests/Azure.Storage.Tests.ps1:127]:
```suggestion
            $ruleResult.Length | Should -Be 1;
            $ruleResult.TargetName | Should -BeIn 'storage-G';
```

**Review by BernieWhite — COMMENTED** [review]
@pazdedav Nice work. Mostly there, just a few suggestions.

**Comment by pazdedav** [comment]
> @pazdedav Nice work. Mostly there, just a few suggestions.

Thanks so much for your suggestions, @BernieWhite . I will need to learn more about Pester for sure ;).

**Review by BernieWhite — APPROVED** [review]
@pazdedav Look great! All good to merge.

**Pull request merged**